### PR TITLE
Don't throw exception out of threadpool

### PR DESCRIPTION
**Pull request by benaadams**
Would like to log with logger in the catches but would need to be with a static var to not allocate. Suggestions?

/cc @davidfowl

### CODE REVIEW

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/TaskUtilities.cs [hunk 1]
```diff
@@ -12,5 +15,72 @@ public static class TaskUtilities
...
+            {
+                // log with a static logger in some way?
+            }
+        };
+
+        private static WaitCallback _completeTcs = (o) =>
+        {
+            try
+            {
+                ((TaskCompletionSource<object>)o).SetResult(null);
```
Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/TaskUtilities.cs]: Was there a situation where this was throwing before? If there was, we could just change this to call TrySetResult which never throws. We can also call TrySetException in ErrorOnThreadPool.

Comment (benaadams) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/TaskUtilities.cs]: No, and I doubt it ever would as this is the only place that ever sets it and it is called only once - however not catching an exception at the top level of a threadpool thread; which is at the top of the stack, will crash the process (by design) as its kinda like `async void`

Comment (benaadams) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/TaskUtilities.cs]: Actually might be better to add a callback to ThreadPool for unhandled execptions

**Comment by benaadams** [comment]
Added https://github.com/dotnet/coreclr/issues/2144 which would probably be easier and better?

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/TaskUtilities.cs (hunk 1 above)
Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/TaskUtilities.cs]: Given that we aren't likely to see the callback added very soon, I think we might want something like this.

The main thing preventing me from merging this is the fact that it doesn't log the exceptions.

**Comment by halter73** [comment]
I think the correct thing to do is make sure that any action we call QueueUserWorkItem with *cannot* throw.

If thre are any places we call QueueUserWorkItem with an action that could possibly throw, we should open another PR that at the very least captures the exception and logs it.

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/TaskUtilities.cs (hunk 1 above)
Comment (benaadams) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/TaskUtilities.cs]: Agreed; though wasn't sure how to get am instance of logger in here? Would setting a static from startup work?

**Comment by benaadams** [comment]
Not sure they are easy places; SocketInput and SocketOutput completions - will do some testing see if its a problem.

**Comment by halter73** [comment]
We can reopen if you strongly disagree. I do see the benefit of having all the try/catch blocks in one place in order to avoid missing stuff, but not at the cost of swallowing exceptions.

Another alternative might be to make the logger a parameter to the `TaskUtilities` methods.

**Comment by benaadams** [comment]
@halter73 updated in  #443 currying to embed the log in closures so its scoped rather than using static

**Pull request closed without merging**